experimental/Schemes/SpaceGerms.jl

```diff
@@ -0,0 +1,233 @@
+
+import AbstractAlgebra: Ring
+
+@Markdown.doc """
+    AbsSpaceGerm{BaseRingType<:Ring,RingType<:Ring}
+
+A space germ, i.e. a ringed space ``(X,O_{(X,x)})`` with local ring ``O_{(X,x)}`` of type `RingType` over a ring ``k`` of type `BaseRingType`.
+"""
+abstract type AbsSpaceGerm{BaseRingType<:Ring, RingType<:Ring} <: AbsSpec{BaseRingType, RingType} end
+
+
+####################################################################
+## two short hand definitions for internal use only
+####################################################################
+GermAtClosedPoint = Spec{<:Ring, 
+                         <:AbsLocalizedRing{<:Ring, <:Any, 
+                                            <:MPolyComplementOfKPointIdeal}
+                        }
+GermAtGeometricPoint = Spec{<:Ring, 
+                            <:AbsLocalizedRing{<:Ring, <:Any, 
+                                               <:MPolyComplementOfPrimeIdeal}
+                           }
+
+###################################################################
+## start of the definition of space germ functionality
+###################################################################
+
+@Markdown.doc """
+    SpaceGerm{BaseRingType, RingType, SpecType}
+A space germ ``(X,O_{(X,x)}``, i.e. a ringed space with underlying scheme ``X`` of type SpecType and local ring ``O_{(X,x)}`` of type `RingType` over some base ring ``k`` of type `BaseRingType`.
+"""
+@attributes mutable struct SpaceGerm{BaseRingType<:Ring, RingType<:Ring, SpecType<:Spec} <: AbsSpaceGerm{BaseRingType, RingType}
+  X::SpecType
+
+  function SpaceGerm(X::GermAtClosedPoint)
+    return new{typeof(base_ring(X)), typeof(OO(X)), typeof(X)}(X)
+  end
+  
+  function SpaceGerm(X::GermAtGeometricPoint)
+    return new{typeof(base_ring(X)), typeof(OO(X)), typeof(X)}(X)
+  end
+end
+
+### Getter functions
+
+function underlying_scheme(X::SpaceGerm)
+  return X.X
+end
+
+@attr Spec function representative(X::SpaceGerm{<:Ring, <:MPolyQuoLocalizedRing})
+    R = base_ring(OO(X))
+    I = modulus(OO(X))
+    Q, _ = quo(R, I)
+    return Spec(Q)
+end
+
+@attr Spec function representative(X::SpaceGerm{<:Ring, <:MPolyLocalizedRing})
+    R = base_ring(OO(X))
+    return Spec(R)
+end
+
+function point(X::SpaceGerm{<:Any, <:Any, <:GermAtClosedPoint})
+  return point_coordinates(inverted_set(OO(X)))
+end
+
+function point(X::SpaceGerm{<:Any, <:Any, <:GermAtGeometricPoint})
+  return prime_ideal(inverted_set(OO(X)))
+end
+
+ring_of_germ(X::AbsSpaceGerm) = OO(X)
+
+function ideal_of_germ(X::AbsSpaceGerm{<:Ring,<:MPolyQuoLocalizedRing})
+    return localized_ring(OO(X))(modulus(OO(X)))
+end
+
+function ideal_of_germ(X::AbsSpaceGerm{<:Ring,<:MPolyLocalizedRing})
+    return ideal(OO(X),[zero(OO(X))])
+end
+
+function ambient_germ(X::AbsSpaceGerm{<:Ring,<:MPolyQuoLocalizedRing})
+    Y,_ = germ_at_point(localized_ring(OO(X)))
+    return Y
+end
+
+function ambient_germ(X::AbsSpaceGerm{<:Ring,<:MPolyLocalizedRing})
```

Comment:
Should this return a copy of `X`?

Reply:
Question: If it is an attribute, it needs to be a copy, right?

Reply:
No. In most places, for efficiency, our getters do not return copies. So if the user mutates them, she has to be careful.

Reply:
Sorry, I just accidentally commited exactly that change. In this case, I vote for a copy, because it really is a different object and one would like to be able to store attributes in it, for instance. Say the user wants a green ambient scheme, but a red germ. Then once after setting the ambient scheme to green, the germ turned green as well? That would be a mess! Let's hand over a copy.

Reply:
Well mutating something is usually unsafe. If the user wants a green ambient scheme `green(::X)` should copy `X` and then paint it green.  (and with `green!(X)` I wouldn't be surprised if anything contained in `X` turns green too and some stuff remains red because we/she forgot about something.) See 
https://oscar-system.github.io/Oscar.jl/dev/DeveloperDocumentation/debugging/#Pitfalls:-Mutable-objects-in-Oscar-code
for more on this.

Reply:
There are two aspects:
- for a subgerm X,x inside A^n,x we need to see A^n,x here, as is the content of the variant which is the heart of this command
- in the setting in question, we are handling A^n,x of which the ambient space is just A^n,x -- but for compatibility on the user-facing level, ambient_germ should be available (for easier scripting....)
Now the question is: should these two A^n,x be the same or different objects? If they are different, we cannot compare whether two germs have the same ambient germ (if one of them happens to be A^n), if however the user decides to do changes to the orignal germ and we have not made a copy, the content of this attribute of the germ might become nonsense.

Reply:
I am strongly against `green!`. Mutating objects should be avoided unless strictly necessary. 
And so I think that we should not allow mutations as part of the interface. Now the user can still 
paint the ambient space green if the likes - but at her own risk - we guarantee for nothing then. 

Here is a somewhat stupid example: 
```
julia> P,_ = PolynomialRing(GF(2),[:x,:y])
(Multivariate Polynomial Ring in x, y over Galois field with characteristic 2, gfp_mpoly[x, y])

julia> B = base_ring(P)  # not a copy
Galois field with characteristic 2

julia> B.n= 5 # paint it green
5

julia> P
Multivariate Polynomial Ring in x, y over Galois field with characteristic 5
```
Now we cannot expect that anything defined over `P` makes sense anymore - but we did it at our own risk. 
Now if there is no `green!` function it is not necessary to return a copy.

Reply:
Here is another less constructed example:
```
julia> P, (x,y) = PolynomialRing(QQ,[:x,:y])
(Multivariate Polynomial Ring in x, y over Rational Field, fmpq_mpoly[x, y])

julia> I = ideal(P,[x,y])
ideal(x, y)

julia> add!(gens(I)[1],x, y)  # mutate a generator
x + y

julia> I
ideal(x + y, y)
```
Therefore the user can at no point assume that `green!` is safe to use.

Reply:
@HechtiDerLachs @simonbrandhorst: We probably won't resolve this point without a face to face discussion. There are usability issues to be considered and we will eventually have to weigh the pros and cons for at least three possible ways to handle this. Maybe we could meet online tomorrow or on wednesday.